src/collapse/dtrace.rs

```diff
@@ -0,0 +1,214 @@
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        for _ in 0..HEADER_LINES {
+            reader.read_line(&mut line)?;
+        }
+
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Self {
+            stack: VecDeque::default(),
+            occurrences: HashMap::default(),
+            opt: opt,
+            stack_str_size: 0,
+        }
+    }
+}
+
+impl Folder {
+    fn remove_offset(line: &str) -> &str {
+        let mut line = line.rsplitn(2, '+');
+        // at least one element will be returned
+        let second = line.next().unwrap();
+        if let Some(first) = line.next() {
+            first
+        } else {
+            second
+        }
+    }
+
+    // we have a stack line that shows one stack entry from the preceeding event, like:
+    //
+    //     unix`tsc_gethrtimeunscaled+0x21
+    //     genunix`gethrtime_unscaled+0xa
+    //     genunix`syscall_mstate+0x5d
+    //     unix`sys_syscall+0x10e
+    //       1
+    fn on_stack_line(&mut self, line: &str) {
+        let line = line.trim_start();
+        let frame = if self.opt.includeoffset {
+            line
+        } else {
+            Self::remove_offset(line)
+        };
+
+        let mut frame = CPP_REGEX.replace_all(frame, "$n").to_string();
+        let mut inline = false;
+
+        if frame.is_empty() {
+            frame = "-".to_string();
+        };
+
+        for func in frame.split("->") {
+            let mut func = func.replace(';', ":").trim_start_matches('L').to_string();
+            if inline {
+                func.push_str("_[i]")
+            };
+            inline = true;
+            self.stack_str_size += func.len() + 1;
+            self.stack.push_front(func);
+        }
+    }
+
+    fn on_stack_end(&mut self, count: usize) {
+        // allocate a string that is long enough to hold the entire stack string
+        let mut stack_str = String::with_capacity(self.stack_str_size);
+
+        let mut first = true;
+        // add the other stack entries (if any)
+        for e in self.stack.drain(..) {
+            if first {
+                first = false
+            } else {
+                stack_str.push_str(";");
+            }
+            stack_str.push_str(&e);
+        }
+        //trim leaf offset if these were retained:
+        if self.opt.includeoffset {
+            stack_str = Self::remove_offset(&stack_str).to_string();
```

Comment:
It'd be great if we could find a way to not re-allocate a `String` here since we already have a `String`.

Reply:
I moved this inside the loop, that way it can remain a &str and at the same time operates over a smaller string so a win win :)